bots/nymphora/nymphora_main.lua

```diff
@@ -0,0 +1,613 @@
+object.nPuzzleThreshold = 30
+
+-- Ability up bonuses --
+object.nStunUp = 10
+object.nHealUp = 5
+
+object.nSheepstickUp = 15
+object.nFrostfieldUp = 5
+
+--------------------
+-- For skillbuild --
+--------------------
+object.nManaNeeded = 0
+object.nHealNeeded = 0
+
+--------------
+-- For heal --
+--------------
+object.vecHealPos = nil
+object.nHealLastCastTime = -20000
+
+
+local function AbilitiesUpUtility(hero)
+	local nUtility = 0
+	
+	if skills.heal:CanActivate() then
+		nUtility = nUtility + object.nHealUp
+	end
+	
+	if skills.stun:CanActivate() then
+		nUtility = nUtility + object.nStunUp
+	end
+	
+	local itemSheepstick = core.GetItem("Item_Morph")
+	if itemSheepstick ~= nil and itemSheepstick:CanActivate() then
+		nUtility = nUtility + object.nSheepstickUp
+	end
+
+	local itemFrostfieldPlate = core.GetItem("Item_FrostfieldPlate")
+	if itemFrostfieldPlate ~= nil and itemFrostfieldPlate:CanActivate() then
+		nUtility = nUtility + object.nFrostfieldUp
+	end
+	
+	return nUtility
+end
+
+function object:oncombateventOverride(EventData)
+	self:oncombateventOld(EventData)
+
+	local nAddBonus = 0
+	
+	if EventData.Type == "Ability" then
+		-- No use bonuses. Add up bonuses to not drop aggro as skills are no longer 'up'
+		if EventData.InflictorName == "Ability_Fairy1" then
+			nAddBonus = nAddBonus + object.nHealUp
+			object.nGraveyardUseTime = EventData.TimeStamp
+		elseif EventData.InflictorName == "Ability_Fairy3" then
+			nAddBonus = nAddBonus + object.nHealUp
+		end
+	elseif EventData.Type == "Item" then
+		if EventData.SourceUnit == core.unitSelf:GetUniqueID() then
+			if EventData.InflictorName == "Item_Morph" then
+				nAddBonus = nAddBonus + object.nSheepstickUp
+			elseif EventData.InflictorName == "Item_FrostfieldPlate" then
+				nAddBonus = nAddBonus + object.nFrostfieldUp
+			end
+		end
+	end
+	
+	if nAddBonus > 0 then
+		--decay before we add
+		core.DecayBonus(self)
+	
+		core.nHarassBonus = core.nHarassBonus + nAddBonus
+	end
+end
+object.oncombateventOld = object.oncombatevent
+object.oncombatevent 	= object.oncombateventOverride
+
+function object:SkillBuild()
+	local unitSelf = self.core.unitSelf
+
+	if skills.heal == nil then
+		skills.heal		= unitSelf:GetAbility(0)
+		skills.mana		= unitSelf:GetAbility(1)
+		skills.stun		= unitSelf:GetAbility(2)
+		skills.teleport	= unitSelf:GetAbility(3)
+		skills.recall	= unitSelf:GetAbility(5)
+	end
+
+	if unitSelf:GetAbilityPointsAvailable() <= 0 then
+		return
+	end
+
+	for i = 0, unitSelf:GetAbilityPointsAvailable(), 1 do
+		bAbilityLeveled = false
+		if skills.teleport:CanLevelUp() then
+			skills.teleport:LevelUp()
+		else
+			if (self.nHealNeeded > 1 and skills.heal:CanLevelUp()) or (self.nManaNeeded > 1 and skills.mana:CanLevelUp()) or
+				(not skills.stun:CanLevelUp() and (skills.heal:CanLevelUp() or skills.mana:CanLevelUp())) then
+				if self.nHealNeeded > self.nManaNeeded and skills.heal:CanLevelUp() then
+					skills.heal:LevelUp()
+					bAbilityLeveled = true

```

Comment:
bAbilityLeveled is being set to true regardless - so it should be outside the if statement.